libs/partners/ibm/langchain_ibm/llms.py

```diff
@@ -0,0 +1,396 @@
+        return False
+
+    @property
+    def lc_secrets(self) -> Dict[str, str]:
+        return {
+            "url": "WATSONX_URL",
+            "apikey": "WATSONX_APIKEY",
+            "token": "WATSONX_TOKEN",
+            "password": "WATSONX_PASSWORD",
+            "username": "WATSONX_USERNAME",
+            "instance_id": "WATSONX_INSTANCE_ID",
+        }
+
+    @root_validator()
+    def validate_environment(cls, values: Dict) -> Dict:
+        """Validate that credentials and python package exists in environment."""
+        values["url"] = convert_to_secret_str(
+            get_from_dict_or_env(values, "url", "WATSONX_URL")
+        )
+        if "cloud.ibm.com" in values.get("url", "").get_secret_value():
+            values["apikey"] = convert_to_secret_str(
+                get_from_dict_or_env(values, "apikey", "WATSONX_APIKEY")
+            )
+        else:
+            if (
+                not values["token"]
+                and "WATSONX_TOKEN" not in os.environ
+                and not values["password"]
+                and "WATSONX_PASSWORD" not in os.environ
+                and not values["apikey"]
+                and "WATSONX_APIKEY" not in os.environ
+            ):
+                raise ValueError(
+                    "Did not find 'token', 'password' or 'apikey',"
+                    " please add an environment variable"
+                    " `WATSONX_TOKEN`, 'WATSONX_PASSWORD' or 'WATSONX_APIKEY' "
+                    "which contains it,"
+                    " or pass 'token', 'password' or 'apikey'"
+                    " as a named parameter."
+                )
+            elif values["token"] or "WATSONX_TOKEN" in os.environ:
+                values["token"] = convert_to_secret_str(
+                    get_from_dict_or_env(values, "token", "WATSONX_TOKEN")
+                )
+            elif values["password"] or "WATSONX_PASSWORD" in os.environ:
+                values["password"] = convert_to_secret_str(
+                    get_from_dict_or_env(values, "password", "WATSONX_PASSWORD")
+                )
+                values["username"] = convert_to_secret_str(
+                    get_from_dict_or_env(values, "username", "WATSONX_USERNAME")
+                )
+            elif values["apikey"] or "WATSONX_APIKEY" in os.environ:
+                values["apikey"] = convert_to_secret_str(
+                    get_from_dict_or_env(values, "apikey", "WATSONX_APIKEY")
+                )
+                values["username"] = convert_to_secret_str(
+                    get_from_dict_or_env(values, "username", "WATSONX_USERNAME")
+                )
+            if not values["instance_id"] or "WATSONX_INSTANCE_ID" not in os.environ:
+                values["instance_id"] = convert_to_secret_str(
+                    get_from_dict_or_env(values, "instance_id", "WATSONX_INSTANCE_ID")
+                )
+
+        try:
+            from ibm_watsonx_ai.foundation_models import ModelInference  # type: ignore
```

Comment:
You can add this at the top of the file instead of here as a concrete dep!

Reply:
Changed